packages/utils/src/envelope.ts

```diff
@@ -142,40 +134,63 @@ function concatBuffers(buffers: Uint8Array[]): Uint8Array {
   return merged;
 }
 
+function getLineEnd(data: Uint8Array): number {
+  let end = data.indexOf(0xa);
+  if (end === -1) {
+    end = data.length;
+  }
+
+  return end;
+}
+
+function parseJSONFromBuffer(data: Uint8Array): ReturnType<JSON['parse']> {
+  return JSON.parse(new TextDecoder().decode(data));
+}
+
+export type EnvelopeItem = Envelope[1][number];
+
 /**
- * Parses an envelope
+ * Implements parser for
+ * @see https://develop.sentry.dev/sdk/envelopes/#serialization-format
+ * @param rawEvent Envelope data
+ * @returns parsed envelope
  */
 export function parseEnvelope(env: string | Uint8Array): Envelope {
   let buffer = typeof env === 'string' ? encodeUTF8(env) : env;
 
-  function readBinary(length: number): Uint8Array {
-    const bin = buffer.subarray(0, length);
-    // Replace the buffer with the remaining data excluding trailing newline
-    buffer = buffer.subarray(length + 1);
-    return bin;
-  }
-
-  function readJson<T>(): T {
-    let i = buffer.indexOf(0xa);
-    // If we couldn't find a newline, we must have found the end of the buffer
-    if (i < 0) {
-      i = buffer.length;
-    }
-
-    return JSON.parse(decodeUTF8(readBinary(i))) as T;
+  function readLine(length?: number): Uint8Array {
+    const cursor = length != null ? length : getLineEnd(buffer);
+    const line = buffer.subarray(0, cursor);
+    buffer = buffer.subarray(cursor + 1);
+    return line;
   }
 
-  const envelopeHeader = readJson<BaseEnvelopeHeaders>();
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  const items: [any, any][] = [];
+  const envelopeHeader = parseJSONFromBuffer(readLine()) as BaseEnvelopeHeaders;
 
+  const items: EnvelopeItem[] = [];
   while (buffer.length) {
-    const itemHeader = readJson<BaseEnvelopeItemHeaders>();
-    const binaryLength = typeof itemHeader.length === 'number' ? itemHeader.length : undefined;
+    const itemHeader = parseJSONFromBuffer(readLine()) as BaseEnvelopeItemHeaders;
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    const payloadLength = itemHeader.length;
+    let itemPayload = readLine(payloadLength != null ? payloadLength : undefined);
+
+    try {
+      itemPayload = parseJSONFromBuffer(itemPayload);
+    } catch (err) {
+      logger.error(err);
+    }
 
-    items.push([itemHeader, binaryLength ? readBinary(binaryLength) : readJson()]);
+    // data sanitization
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (itemHeader.type && typeof itemPayload === 'object') {
+      /* eslint-disable @typescript-eslint/no-unsafe-member-access */
+      // @ts-expect-error -- Does not like assigning to `type` on random object
+      itemPayload.type = itemHeader.type;
+    }
```

Comment:
The root of all the test changes with `type: 'event'` or `type: 'span'` etc is this. I assumed Spotlight was the correct one here and all events should have a proper type that said in many places, we assume error events to not have any type.

I feel like an explicit type is important so even if this implementation is not correct, we should have a default event type like `'event'` or `'error'`.

Anyway, I'll leave this to the reviewers and will address change requests accordingly.

Reply:
h: I wouldn't do this because we use `parseEnvelope` in all our tests to assert on what exactly our SDK sends to Sentry. Adding fields that were not included in the envelope when it was serialized by the SDK would be really confusing.